### PR TITLE
perf(enhance): skip torch._assert_async value checks on MPS (no-kernel CPU fallback syncs per call)

### LINKED ISSUE #3914: ColorJiggle on MPS: rgb_to_hsv max/min-with-indices is ~100x slower than amax/amin; _assert_async CPU fallback adds per-call syncs
## Summary

The augmentation flagship benchmark (#3911, PR #3912) flagged `ColorJiggle` at ~48 img/s on Apple MPS. Profiling on an M-series box (torch 2.9.1, kornia 0.9.0rc1) shows two independent mechanisms, both fixable without changing eager semantics:

## 1. `rgb_to_hsv` uses `max(dim)`/`min(dim)` **with indices** — ~100× slower than `amax/amin` on MPS (the dominant cost)

`kornia/color/hsv.py::rgb_to_hsv` computes `image.max(-3)` and `image.min(-3)` to get channel max/min plus an argmax used only to `gather` one of three hue formulas. On MPS the with-indices reduction over the channel dim is catastrophically slow:

| op (16×3×256×256 f32, MPS) | median |
|---|---|
| `x.max(-3)` (values+indices) | 34.6 ms |
| `x.amax(-3)` (values only) | **0.34 ms** |
| `x.min(-3)` (values+indices) | 34.7 ms — and the argmin is **never used** |
| full `rgb_to_hsv` | 77.6 ms |

`ColorJiggle` pays this twice per call (`adjust_hue` and `adjust_saturation` each round-trip RGB↔HSV): ~160 ms of its ~190 ms per call at batch 16 on MPS. The argmax is only needed to pick 1-of-3 hue sextant formulas, which a branchless `torch.where` chain does with identical first-occurrence tie-breaking — byte-identical outputs and gradients, and pointwise ops fuse under `torch.compile` (the `gather` did not).

Measured with the branchless rewrite: `rgb_to_hsv` **77.8 ms → 8.9 ms MPS (8.7×)**, **39.1 ms → 15.8 ms CPU (2.5×)**.

## 2. `torch._assert_async` has no MPS kernel — per-call CPU fallback sync

`aten::_assert_async.msg` is not implemented on MPS; PyTorch warns and falls back to CPU, forcing a per-call sync (it must materialize the condition, draining the queued stream). Sites: `kornia/enhance/adjust.py` (`adjust_gamma` ×2, `adjust_contrast`, `solarize`).

Standalone `adjust_contrast` (16×3×256×256, MPS): **5.83 ms with the assert vs 1.73 ms without (3.4×)**. Inside `ColorJiggle` it is only ~2% of the end-to-end time (the hsv cost above dominates), so this is the smaller of the two fixes — the original attribution of the ColorJiggle slowness to this assert was wrong.

## Fix plan (two small PRs)

1. **perf(color): branchless hue selection in `rgb_to_hsv`** — `amax`/`amin` + `where`-chain replicating `torch.max(dim)` tie-breaking (gray/tie pixels pinned by a new test). Byte-identical outputs and gradients vs current implementation on CPU/MPS for f16/f32/f64, verified incl. channel-tie inputs; still TorchScript-scriptable and `fullgraph=True`-compilable (1 graph, 0 breaks).
2. **perf(enhance): make value asserts MPS-safe** — keep `torch._assert_async` where it has a native kernel (CPU/CUDA), skip it on MPS where the only alternative is a per-call sync, and upgrade the messages to state the fix (W4).

Related: the compiled `ColorJiggle` failure reported in the benchmark findings is **not** the hsv path — after fix 1 it still fails with `GuardOnDataDependentSymNode` at the random-order `transforms[idx]` dispatch, i.e. the compile-safe parameter-generator work tracked in #3913.

Hardware/software: Apple M-series (arm64 macOS), torch 2.9.1, kornia 0.9.0rc1, `torch.utils.benchmark` medians.

### PR DESCRIPTION
**Pull request by ducha-aiki**
Fixes #3914 (mechanism 2 of 2; mechanism 1 is #3915).

## What

`aten::_assert_async` has no MPS kernel. PyTorch's CPU fallback materializes the condition tensor, which drains the queued MPS stream on **every call** — a hidden full device sync (plus a `UserWarning` per site). The three affected sites (`adjust_gamma` ×2, `adjust_contrast`, `solarize` in `kornia/enhance/adjust.py`) now route through a private `_assert_async_value_check` helper that keeps the fullgraph-safe `torch._assert_async` on CPU/CUDA and skips the check on MPS, where the only way to evaluate it is to sync. Messages upgraded to state the fix (W4, e.g. "Contrast factor must be non-negative. Sample it from a non-negative range or clamp: factor.clamp_min(0.0).").

Behavior change is confined to **invalid inputs on MPS**: they no longer raise (there is no async assert to raise them without a per-call sync); valid-input outputs are untouched on every backend. Both behaviors are pinned by new tests (`test_negative_factor_errors` on CPU, `test_mps_skips_value_check` on MPS), written first and watched fail (TDD: the MPS one fails on `main` with `RuntimeError` from the CPU fallback).

## Compile

`torch.compile(adjust_contrast, fullgraph=True)`: 1 graph, 0 breaks, matches eager — the `cond.device.type` branch resolves at trace time (static device metadata), so eager and compiled run the same path per device; this is not an `is_compiling` split. `tests/enhance/test_adjust.py -k dynamo` under `KORNIA_TEST_OPTIMIZER=inductor`: 7 passed, 1 skipped.

## Benchmark (op level)

Apple M-series, torch 2.9.1, `torch.utils.benchmark` medians, 16×3×256×256 f32 on MPS, tensor-valued factors, same box back-to-back (before = `d44abd15`):

| op (MPS) | main | this PR | speedup |
|---|---|---|---|
| `adjust_contrast` | 4.87 ms | 2.46 ms | **2.0×** |
| `adjust_gamma` (2 asserts) | 7.03 ms | 5.08 ms | **1.4×** |
| `solarize` | 7.79 ms | 4.36 ms | **1.8×** |

CPU/CUDA are unaffected by construction (same `torch._assert_async` call as before; the helper adds one Python call and a device-type string compare, below measurement noise). Within `ColorJiggle` this is the smaller mechanism (~2% end-to-end — the dominant one was `rgb_to_hsv`, #3915); standalone `adjust_contrast`/`solarize` users on MPS get the full win, and the per-call `aten::_assert_async` fallback `UserWarning` spam is gone.

## Algorithm source

No algorithm change — validation plumbing only; the guarded-assert pattern follows the existing fullgraph-safe convention already in this file.

## Test log

```
tests/enhance/test_adjust.py --dtype=all --device=cpu   → 454 passed, 25 skipped
tests/enhance/test_adjust.py --dtype=float32,float16 --device=mps
                                                        → 216 passed, 29 skipped, 4 failed*
KORNIA_TEST_OPTIMIZER=inductor ... -k dynamo            → 7 passed, 1 skipped
```

\* The 4 MPS failures are `TestPosterize::test_cardinality[mps-float16-…]` device-mismatch errors that reproduce identically with `main`'s `kornia/enhance/adjust.py` (verified by flipping only that file back to `d44abd15`) — pre-existing `posterize`-on-MPS issue, untouched by this PR.

---

Posted on behalf of @ducha-aiki by Claude (Fable 5).

### CODE REVIEW

File: tests/enhance/test_adjust.py [hunk 1]
```diff
@@ -359,6 +359,25 @@ def test_factor_one_acumulative(self, device, dtype):
         f = kornia.enhance.AdjustContrast(1.0)
         self.assert_close(f(data), expected)
 
+    def test_negative_factor_errors(self, device, dtype):
+        if device.type != "cpu":
+            pytest.skip("value asserts are synchronous only on CPU (async on CUDA, skipped on MPS)")
+        img = torch.rand(1, 3, 4, 4, device=device, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply tests/enhance/test_adjust.py:365]: Declined for file consistency: tests/enhance/test_adjust.py (and most of tests/ — 13 of ~200 files carry the future import) deliberately follows the BaseTester style without it; adding it only for this diff would be inconsistent with the module. Same rationale as the declined suggestion on the #3912 benchmark utils.

File: tests/enhance/test_adjust.py [hunk 2]
```diff
@@ -359,6 +359,25 @@ def test_factor_one_acumulative(self, device, dtype):
...
+        factor = torch.tensor([-0.5], device=device, dtype=dtype)
+        with pytest.raises(RuntimeError, match="non-negative"):
+            kornia.enhance.adjust_contrast(img, factor)
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply tests/enhance/test_adjust.py:368]: Declined: every existing test method in this file uses untyped (self, device, dtype) fixture signatures; typing only the two new methods would be inconsistent with the file. Precedent: identical suggestion declined on #3912.

File: kornia/enhance/adjust.py [hunk 3]
```diff
@@ -292,10 +305,8 @@ def adjust_gamma(
     gamma = gamma.to(input.device).to(input.dtype)
     gain = gain.to(input.device).to(input.dtype)
 
-    # torch._assert_async keeps the value check while staying fullgraph-compilable (a Python
-    # `if tensor: raise` would break the graph).
-    torch._assert_async((gamma >= 0.0).all(), "Gamma must be non-negative.")
-    torch._assert_async((gain >= 0.0).all(), "Gain must be non-negative.")
+    _assert_async_value_check((gamma >= 0.0).all(), "Gamma must be non-negative. Clamp it first: gamma.clamp_min(0.0).")
+    _assert_async_value_check((gain >= 0.0).all(), "Gain must be non-negative. Clamp it first: gain.clamp_min(0.0).")
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/enhance/adjust.py]: Checked with awk 'length > 120': the only such lines in kornia/enhance/adjust.py are 340/474/1019 — pre-existing docstring URLs untouched by this PR. All added lines are formatted by the pinned ruff (v0.15.1) and pre-commit.ci passes on this PR; the cited ranges are within the limit as of a89c45fa.

[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/enhance/adjust.py]: Reworded in a89c45fa: each message now gives both forms — max(x, 0.0) for float callers and x.clamp_min(0.0) (resp. additions.clamp(-0.49, 0.49)) for tensor callers.

File: tests/enhance/test_adjust.py [hunk 4]
```diff
@@ -359,6 +359,25 @@ def test_factor_one_acumulative(self, device, dtype):
...
+    def test_mps_skips_value_check(self, device, dtype):
+        # aten::_assert_async has no MPS kernel; its CPU fallback forces a full device sync per
+        # call. On MPS the value check is skipped entirely (documented in adjust_contrast), so
+        # even an invalid factor must go through without raising or falling back to CPU.
+        if device.type != "mps":
+            pytest.skip("pins the MPS-only no-CPU-fallback behavior")
+        img = torch.rand(1, 3, 4, 4, device=device, dtype=dtype)
+        factor = torch.tensor([-0.5], device=device, dtype=dtype)
+        out = kornia.enhance.adjust_contrast(img, factor)
+        assert out.shape == img.shape
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply tests/enhance/test_adjust.py:380]: Added out.device and out.dtype assertions in a89c45fa.

File: tests/enhance/test_adjust.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply tests/enhance/test_adjust.py:368]: Intentionally CPU-only: on CUDA a failing torch._assert_async raises a device-side assert, which leaves the CUDA context poisoned for every subsequent test in the process — synchronizing inside pytest.raises catches the error but does not un-poison the context. This async-fault stickiness is exactly what corrupted the L4 benchmark runs tracked in #3913, so the raise-check stays CPU-scoped.

File: tests/enhance/test_adjust.py [hunk 5]
```diff
@@ -359,6 +359,25 @@ def test_factor_one_acumulative(self, device, dtype):
...
+            pytest.skip("value asserts are synchronous only on CPU (async on CUDA, skipped on MPS)")
+        img = torch.rand(1, 3, 4, 4, device=device, dtype=dtype)
+        factor = torch.tensor([-0.5], device=device, dtype=dtype)
+        with pytest.raises(RuntimeError, match="non-negative"):
+            kornia.enhance.adjust_contrast(img, factor)
+
+    def test_mps_skips_value_check(self, device, dtype):
+        # aten::_assert_async has no MPS kernel; its CPU fallback forces a full device sync per
+        # call. On MPS the value check is skipped entirely (documented in adjust_contrast), so
+        # even an invalid factor must go through without raising or falling back to CPU.
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply tests/enhance/test_adjust.py]: The referenced note now exists: a89c45fa adds it to the adjust_contrast docstring (and adjust_gamma/solarize), and the test comment points at it explicitly.

File: tests/enhance/test_adjust.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply tests/enhance/test_adjust.py:380]: Added out.device and out.dtype assertions in a89c45fa.

File: kornia/enhance/adjust.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/enhance/adjust.py]: Documented in a89c45fa: adjust_gamma's docstring now notes the check runs on CPU/CUDA via torch._assert_async and is skipped on MPS (no kernel; CPU fallback would sync per call).

File: kornia/enhance/adjust.py [hunk 6]
```diff
@@ -368,8 +379,10 @@ def adjust_contrast(image: torch.Tensor, factor: Union[float, torch.Tensor], cli
     while len(factor.shape) != len(image.shape):
         factor = factor[..., None]
 
-    # torch._assert_async keeps the value check while staying fullgraph-compilable.
-    torch._assert_async((factor >= 0).all(), "Contrast factor must be positive.")
+    _assert_async_value_check(
+        (factor >= 0).all(),
+        "Contrast factor must be non-negative. Sample it from a non-negative range or clamp: factor.clamp_min(0.0).",
+    )
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/enhance/adjust.py:402]: Documented in a89c45fa: adjust_contrast's docstring now carries the note (same wording as adjust_gamma/solarize).

File: kornia/enhance/adjust.py [hunk 7]
```diff
@@ -722,11 +735,9 @@ def solarize(
...
             additions = torch.as_tensor(additions)
 
-        # `torch._assert_async` keeps this a single traceable path (no Python-level branch on a
-        # tensor value), so `solarize` stays torch.compile fullgraph-safe while still validating.
-        torch._assert_async(
+        _assert_async_value_check(
             ((additions < 0.5) & (additions > -0.5)).all(),
-            "The value of 'addition' is between -0.5 and 0.5.",
+            "The addition must be in the open range (-0.5, 0.5). Clamp it first, e.g. additions.clamp(-0.49, 0.49).",
         )
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/enhance/adjust.py:764]: Documented in a89c45fa: solarize's docstring now notes the additions range check runs on CPU/CUDA and is skipped on MPS, with the reason.
